### PR TITLE
[ fix #256 ] Liberate AVL trees from propositional equality

### LINKED ISSUE #256: Using String as a key in an AVL tree
Hi, I'd like to use `String` as a key in an AVL tree. `String` has an instance of `StrictTotalOrder`, but it still can't be used because the underlying notion of equality isn't Agda's builtin. When I try this:
```
primitives : Tree {Key = String}            -- Key
                  (λ _ → List Lisp → Lisp)  -- Value
                  (isStrictTotalOrder String.strictTotalOrder)
```
I get this:
```
.Relation.Binary.List.Pointwise.Rel
(_≈_ .Data.Char.strictTotalOrder) (String.toList x)
(String.toList x)
!= x .Agda.Builtin.Equality.≡ x of type Set
when checking that the expression
isStrictTotalOrder String.strictTotalOrder has type
Relation.Binary.IsStrictTotalOrder .Agda.Builtin.Equality._≡_
__<__9
```
The following also doesn't work for `List Char`:
```
 (isStrictTotalOrder (StrictLex.<-strictTotalOrder Char.strictTotalOrder))
```
I get this message:
```
/home/siddharthist/code/agda-scheme/Eval.agda:20,9-80
.Relation.Binary.List.Pointwise.Rel (_≈_ Char.strictTotalOrder) x x
!= x .Agda.Builtin.Equality.≡ x of type Set
when checking that the expression
isStrictTotalOrder (<-strictTotalOrder Char.strictTotalOrder) has
type
Relation.Binary.IsStrictTotalOrder .Agda.Builtin.Equality._≡_
__<__10
``` 
Indeed, one can't even use `Char.strictTotalOrder`, which relies on equality of `Nat`s. Could the condition on the underlying equivalence relation be abandoned?

I have Agda version 2.5.3, and am testing with commit 7220ca3ee025e9339aa8e5d11efcd069764e7977.

### PR DESCRIPTION
**Pull request by gallais**
joint work with @fredrikNordvallForsberg

* The family of Values stored in the tree now needs to respect the
  Key equivalence

* We have noticed that Data.AVL.Indexed's lookup & delete didn't use
  a range to guarantee that the recursive calls were performed in the
  right subtree. The types have been made more precise.

* (insert/union)With functions now that a function `Maybe Val -> Val`
  rather than a value together with a merging function `Val -> Val -> Val`
  to handle the case where a value is already present at that key.

* Various functions have been made polymorphic which makes their biases
  & limitations clearer. e.g. we have:
  unionWith : (V -> Maybe W -> W) -> Tree V -> Tree W -> Tree W
  but ideally we would like to have:
  unionWith : (These V W -> X) -> Tree V -> Tree W -> Tree X

### CODE REVIEW

**Comment by gallais** [comment]
After the inclusion of `alignWith` in `Data.Maybe.Base`, we get a cyclic dependency
with `Data.These` because of these:

```agda
-- Injections.

  thisM : A → Maybe B → These A B
  thisM a = maybe′ (these a) (this a)

  thatM : Maybe A → B → These A B
  thatM = maybe′ these that
```

One solution would be to move the `align` function from `Data.Maybe.Base` to `Data.Maybe` in #464

**Comment by MatthewDaggitt** [comment]
Hmm if we did that, then we should do it consistently across all data...

**Comment by gallais** [comment]
The rest of this PR doesn't use `thisM` and `thatM` so we could also forget about them.
(Or put them in `Data.Maybe` in a different PR).

**Comment by gallais** [comment]
> Hmm if we did that, then we should do it consistently across all data...

Btw in #447 I have introduced `Data.These.Base` and moved all the Maybe-independent
definitions of `Data.These` into it to solve a similar dependency cycle problem.

I think this is the way to go. Probably worth scheduling this one after #447.

**Comment by MatthewDaggitt** [comment]
Just looked through this.

> Btw in #447 I have introduced Data.These.Base and moved all the Maybe-independent
definitions of Data.These into it to solve a similar dependency cycle problem.

I think the actual problem is that there's bidirectional dependencies between `Data.These` and `Data.Maybe`. Every other type fits neatly into a partial order of dependencies. From the look of it `Data.These` is more "fundemental" than `Data.Maybe`. Therefore I've moved `thisM` and `thatM` to `Data.Maybe.Base` which neatly solves the problem.

Also now that we're not having to enforce using propositional equality ideally the interface should use a `StrictTotalOrder` rather than an `IsStrictTotalOrder`...  Have we broken the interface already or is it still technically backwards compatible?

**Comment by gallais** [comment]
> Have we broken the interface already or is it still technically backwards compatible?

I have changed in a non-backwards compatible way the types of `(union/inser)(/With)`.
The introduction of `src/Data/AVL/Value.agda` also breaks backwards compatibility.

So I think we should indeed use a `StrictTotalOrder`.

File: src/Data/AVL/Value.agda [hunk 1]
```diff
@@ -5,19 +5,22 @@
...
+    family   : Key → Set v
+    respects : family Respects _≈_
 
 const : ∀ {v} → Set v → Value v
-Value.family   (const V) = F.const V
-Value.respects (const V) = F.const F.id
+const V = record
+  { family   = F.const V
+  ; respects = F.const F.id
+  }
```
[the remark this answers is not in the document]
Comment (MatthewDaggitt) [reply src/Data/AVL/Value.agda]: Ah apologies, I hadn't come across that feature before! I'll change it back and leave a note explaining why.

**Pull request merged**